tests/functions.py

```diff
@@ -227,19 +227,62 @@ def authenticate(dataset):
         )
 
 
-def get_filenames(dataset):
-    annex_list: str = git.Repo(dataset).git.annex("list")
-    filenames: List[str] = re.split(r"\n[_X]+\s", annex_list)[1:]
-    return filenames
+def get_filenames(dataset, *, minimum):
+    contains_archived_files = False
+    annex_list: str = iter(git.Repo(dataset).git.annex("list").split("\n"))
+    remotes = list()
+
+    # Retrieve remotes from the header.
+    for line in annex_list:
+        if re.match(r"^\|+$", line):
+            break
+        remotes.append(re.sub(r"^\|*", "", line))
+
+    if "datalad-archives" in remotes:
+        archived_files = list()
+        independent_files = list()
+        archive_index = remotes.index("datalad-archives")
+
+        for line in annex_list:
+            # Split only for first occurence to prevent failure when filename has spaces.
+            in_remote, filename = line.split(" ", maxsplit=1)
+
+            if in_remote[archive_index] == "X":
+                archived_files.append(filename)
+            else:
+                independent_files.append(filename)
+
+        if len(independent_files) > minimum:
+            filenames = independent_files
+        else:
+            contains_archived_files = True
+            filenames = archived_files + independent_files
 
+    else:
+        filenames = [x.split()[1] for x in annex_list]
+
+    return filenames, contains_archived_files
+
+
+def download_files(dataset, dataset_size, *, num=4):
```

Comment:
@mathdugre shouldn't the `*` be in front of an argument? like `*arg`?

Reply:
@cmadjar the `*` signifies that further arguments are keyword-only. This enforce some arguments to be specified with their keyword.

For example:
`download_files("project/prevent-ad", 12345, 5)` would fail
 `download_files("project/prevent-ad", 12345, num=5)` is fine.


You can find more details about keyword-only arguments in [PEP 3102](https://www.python.org/dev/peps/pep-3102/).

Reply:
Thank you @mathdugre !